programs/clearing_house/src/lib.rs

```diff
@@ -2619,6 +2712,247 @@ pub mod clearing_house {
+            last_withdraw_request_value: 0,
+            last_withdraw_request_ts: 0,
+            cost_basis: 0,
+            expo: 0,
+            last_valid_ts: now,
+        };
+
+        Ok(())
+    }
+
+    pub fn settle_bank_to_insurance_fund<'info>(
+        ctx: Context<SettleBankToInsuranceFund>,
+        bank_index: u64,
+    ) -> Result<()> {
+        let bank = &mut load_mut!(ctx.accounts.bank)?;
+
+        let bank_vault_amount = ctx.accounts.bank_vault.amount;
+        let insurance_vault_amount = ctx.accounts.insurance_fund_vault.amount;
+
+        let clock = Clock::get()?;
+        let now = clock.unix_timestamp;
+
+        let token_amount = controller::insurance::settle_bank_to_insurance_fund(
+            bank_vault_amount,
+            insurance_vault_amount,
+            bank,
+            now,
+        )?;
+
+        controller::token::send_from_bank_vault(
+            &ctx.accounts.token_program,
+            &ctx.accounts.bank_vault,
+            &ctx.accounts.insurance_fund_vault,
+            &ctx.accounts.bank_vault_authority,
+            bank_index,
+            bank.vault_authority_nonce,
+            token_amount as u64,
+        )?;
+
+        Ok(())
+    }
+
+    pub fn add_insurance_fund_stake<'info>(
+        ctx: Context<AddInsuranceFundStake>,
+        bank_index: u64,
+        amount: u64,
+    ) -> Result<()> {
+        if amount == 0 {
+            return Err(ErrorCode::InsufficientDeposit.into());
+        }
+
+        let clock = Clock::get()?;
+        let remaining_accounts_iter = &mut ctx.remaining_accounts.iter().peekable();
+
+        let insurance_fund_stake = &mut load_mut!(ctx.accounts.insurance_fund_stake)?;
+        let user_stats = &mut load_mut!(ctx.accounts.user_stats)?;
+
+        validate!(
+            insurance_fund_stake.bank_index == bank_index,
+            ErrorCode::DefaultError,
+            "insurance_fund_stake does not match bank_index"
+        )?;
+
+        validate!(
+            insurance_fund_stake.last_withdraw_request_shares == 0
+                && insurance_fund_stake.last_withdraw_request_value == 0,
+            ErrorCode::DefaultError,
+            "withdraw request in progress"
+        )?;
+
+        let _oracle_map = OracleMap::load(remaining_accounts_iter, clock.slot)?;
+        let bank_map = BankMap::load(&get_writable_banks(bank_index), remaining_accounts_iter)?;
+
+        controller::insurance::add_insurance_fund_stake(
+            amount,
+            ctx.accounts.insurance_fund_vault.amount,
+            insurance_fund_stake,
+            user_stats,
+            &mut *bank_map.get_ref_mut(&bank_index)?,
+        )?;
+
+        controller::token::receive(
+            &ctx.accounts.token_program,
+            &ctx.accounts.user_token_account,
+            &ctx.accounts.insurance_fund_vault,
+            &ctx.accounts.authority,
+            amount,
+        )?;
+
+        Ok(())
+    }
+
+    pub fn request_remove_insurance_fund_stake<'info>(
```

Comment:
might be confusing ux since you passing in n_shares you want to unstake and it would change function
might be better to be forced to rebase via instruction if its needed before submitting withdraw req?